### PR TITLE
debug logging print only unprocessed segments detail

### PR DESCRIPTION
**Pull request by wangzlei**
*Issue #, if available:*
https://github.com/aws/aws-xray-daemon/issues/63

*Description of changes:*
Only log unprocessed segments by comparing the IDs.
Here because batches is a string slice, filter the unprocessed segments by matching string 'id' will get complexity O(m*n). 
To reduce the complexity we create a map for batches.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by wangzlei** [comment]
*Test done*
```
2020-08-13T21:39:03-07:00 [Info] Sent batch of 4 segments but had 2 Unprocessed segments (0.084 seconds)
2020-08-13T21:39:03-07:00 [Error] Unprocessed segment: {
  ErrorCode: "InvalidId",
  Id: "1835",
  Message: "Invalid segment. ErrorCode: InvalidId"
}
2020-08-13T21:39:03-07:00 [Debug] {"name":"Test1","id":"1835","start_time":1.597379941754E9,"trace_id":"1-5f361565-45bca2da747ffc14a18f43ec","end_time":1.597379942259E9,"subsegments":[{"name":"Test1Sub","id":"0cf0b254d3fdcabd","start_time":1.597379942259E9,"end_time":1.597379942259E9}],"aws":{"xray":{"sdk_version":"2.4.0","sdk":"X-Ray for Java"}},"service":{"runtime":"Java HotSpot(TM) 64-Bit Server VM","runtime_version":"11.0.5"}}
2020-08-13T21:39:03-07:00 [Error] Unprocessed segment: {
  ErrorCode: "InvalidId",
  Id: "1836",
  Message: "Invalid segment. ErrorCode: InvalidId"
}
2020-08-13T21:39:03-07:00 [Debug] {"name":"Test1","id":"1836","start_time":1.59737994226E9,"trace_id":"1-5f361566-501841d9848dc1a28fc5a875","end_time":1.597379942764E9,"subsegments":[{"name":"Test1Sub","id":"1194eaddc9dd0066","start_time":1.597379942764E9,"end_time":1.597379942764E9}],"aws":{"xray":{"sdk_version":"2.4.0","sdk":"X-Ray for Java"}},"service":{"runtime":"Java HotSpot(TM) 64-Bit Server VM","runtime_version":"11.0.5"}}
```

File: pkg/processor/batchprocessor.go [hunk 1]
```diff
@@ -96,10 +97,11 @@ func (s *segmentsBatch) poll() {
 				for _, unprocessedSegment := range r.UnprocessedTraceSegments {
 					telemetry.T.SegmentRejected(1)
 					log.Errorf("Unprocessed segment: %v", unprocessedSegment)
-				}
-				log.Debug("Batch that contains unprocessed segments")
-				for i := 0; i < len(batch); i++ {
-					log.Debug(*batch[i])
+					for i := 0; i < len(batch); i++ {
```
Comment (anuraaga) [inline pkg/processor/batchprocessor.go:100]: Ideally we'd wrap this with something like `isDebugEnabled` but I can't find anything like that in this log library. Actually, this log library is very old and seems of questionable quality (global lock on every log command, even disabled log levels? Blows my mind https://github.com/cihub/seelog/blob/f561c5e57575bb1e0a2167028b7339b3a8d16fb4/log.go#L208). So probably need to swap it out for zap

Comment (wangzlei) [reply pkg/processor/batchprocessor.go:100]: agree, we'd better change a log lib before merging this code.

**Comment by wangzlei** [comment]
Test done
```
2020-08-18T22:03:00-07:00 [Debug] processor: sending partial batch
2020-08-18T22:03:00-07:00 [Debug] processor: segment batch size: 4. capacity: 50
2020-08-18T22:03:00-07:00 [Info] Sent batch of 4 segments but had 2 Unprocessed segments (0.065 seconds)
2020-08-18T22:03:00-07:00 [Error] Unprocessed segment: {
  ErrorCode: "InvalidId",
  Id: "9467",
  Message: "Invalid segment. ErrorCode: InvalidId"
}
2020-08-18T22:03:00-07:00 [Debug] {"name":"Test1","id":"9467","start_time":1.597813378946E9,"trace_id":"1-5f3cb282-0446651d06513154119a46ba","end_time":1.597813379451E9,"subsegments":[{"name":"Test1Sub","id":"2d062b012c767233","start_time":1.597813379449E9,"end_time":1.597813379449E9}],"aws":{"xray":{"sdk_version":"2.4.0","sdk":"X-Ray for Java"}},"service":{"runtime":"Java HotSpot(TM) 64-Bit Server VM","runtime_version":"11.0.5"}}
2020-08-18T22:03:00-07:00 [Error] Unprocessed segment: {
  ErrorCode: "InvalidId",
  Id: "9468",
  Message: "Invalid segment. ErrorCode: InvalidId"
}
2020-08-18T22:03:00-07:00 [Debug] {"name":"Test1","id":"9468","start_time":1.597813379451E9,"trace_id":"1-5f3cb283-f794276c0ff62d7c013bd6d3","end_time":1.597813379956E9,"subsegments":[{"name":"Test1Sub","id":"7094136a13e80fc3","start_time":1.597813379956E9,"end_time":1.597813379956E9}],"aws":{"xray":{"sdk_version":"2.4.0","sdk":"X-Ray for Java"}},"service":{"runtime":"Java HotSpot(TM) 64-Bit Server VM","runtime_version":"11.0.5"}}
```

File: pkg/processor/batchprocessor.go [hunk 2]
```diff
@@ -90,22 +91,30 @@ func (s *segmentsBatch) poll() {
 			}
 			elapsed := time.Since(start)
 
+			batchesMap := make(map[string]string)
+			idRegexp := regexp.MustCompile(`\"id\":\"(.*?)\"`)
```
Comment (anuraaga) [inline pkg/processor/batchprocessor.go:95]: Make this regexp a constant at the top level, we don't want to compile a regex every time

File: pkg/processor/batchprocessor.go [hunk 3]
```diff
@@ -90,22 +91,30 @@ func (s *segmentsBatch) poll() {
...
+			for i := 0; i < len(batch); i++ {
```
Comment (anuraaga) [inline pkg/processor/batchprocessor.go:96]: Move the mapping logic into the `len() != 0` block

File: pkg/processor/batchprocessor.go (hunk 2 above)
Comment (wangzlei) [reply pkg/processor/batchprocessor.go]: ok

File: pkg/processor/batchprocessor.go (hunk 3 above)
Comment (wangzlei) [reply pkg/processor/batchprocessor.go]: fix in next version

**Pull request merged**